File: src/numericalrecipes/Ch6_SpecialFunctions/betai/betai.py
import numpy as np
import sys
import numericalrecipes as nr


def betai(a, b, x):
    # Returns the incomplete beta function Ix(a, b)
    if ((x < 0) or (x > 1)):
        print('BETAI: bad argument fo x')
        sys.exit()

    # Factor "bt" in front of continued fraction
    if ((x == 0) or (x == 1)):
        # Continued fraction
        if (x < ((a + 1) / (a + b + 2))):
            # ...directly
            return 0
        else:
            # ...after symmetry transformation
            return 1
    else:
        bt = np.exp(nr.gammaln(a + b) - nr.gammaln(a) - nr.gammaln(b) +
                    a * np.log(x) + b * np.log(1 - x))
        # Continued fraction
        if (x < ((a + 1) / (a + b + 2))):
            # ...directly
            return bt * nr.betacf(a, b, x) / a
        else:
            # ...after symmetry transformation
            return 1 - bt * nr.betacf(b, a, 1 - x) / b

File: src/numericalrecipes/Ch6_SpecialFunctions/betai/betai.f90
module mod_betai
   ! Contains the following functions: betai
   use mod_betacf
   use mod_gammaln
   implicit none
   contains

   function betai(a, b, x)
      ! Returns the incomplete beta function Ix(a, b)
      implicit none
      real(8), intent(in)     :: a, b, x
      real(8)                 :: bt, betai

      if ((x < 0.0d0) .or. (x > 1.0d0)) then
         write(*, *) 'BETAI: bad argument for x' 
         stop
      end if

      ! Factor "bt" in front of continued fraction
      if ((x == 0.0d0) .or. (x == 1.0d0)) then
         bt = 0.0d0
      else
         bt = exp(gammaln(a + b) - gammaln(a) - gammaln(b) + &
              a * log(x) + b * log(1.0d0 - x))
      end if

      ! Continued fraction
      if (x < ((a + 1.0d0) / (a + b + 2.0d0))) then
         ! ...directly
         betai = bt * betacf(a, b, x) / a
         return
      else
         ! ...after symmetry transformation
         betai = 1.0d0 - bt * betacf(b, a, 1.0d0 - x) / b
         return
      end if

   end function betai

end module mod_betai